azure/services/agentpools/spec.go

```diff
@@ -148,6 +152,7 @@ func (s *AgentPoolSpec) Parameters(existing interface{}) (params interface{}, er
 				MaxCount:            existingPool.MaxCount,
 				NodeLabels:          existingPool.NodeLabels,
 				NodeTaints:          existingPool.NodeTaints,
+				Tags:                existingPool.Tags,
```

Comment:
@nojnhuh @zioproto 

I think this is what may be causing the throttling issue

Looking at https://storage.googleapis.com/kubernetes-jenkins/pr-logs/pull/kubernetes-sigs_cluster-api-provider-azure/2745/pull-cluster-api-provider-azure-e2e-exp/1587117062692540416/artifacts/clusters/capz-e2e-eju8ih-aks/azure-activity-logs/capz-e2e-eju8ih-aks.log, there hundreds (!) of agentpool requests, which is unexpected.

I started looking at what could be causing this and I believe it's because we're comparing all the existing tags on the agentpool with our spec tags, which is always going to cause a diff since AKS adds built in tags. Because of this, the agentpools are getting updated non-stop on every loop, and eventually causing throttling, which causes the e2e test to fail.

Reply:
I looked into this but I don't understand the "AKS adds built in tags".

I tried to create an AKS cluster with minimal options:

```
az aks create -g debug -n myAKSCluster
az aks list -o table
az aks nodepool list --cluster-name myAKSCluster -g debug -o json | jq '.[].tags'
```
and the NodePool does not have any default tag.

I also tried creating a cluster with Cluster API with the stable version and I dont see any tag associated to the Node Pools.

Reply:
Could you add a log statement in here so we can see what diff is triggering the updates in the e2e test logs?

Note: #2611 is being worked on which will help with things like this in the future, but let's add a temporary log here so we can understand what's going on

Reply:
I added this ugly patch in my local env:
```
diff --git a/azure/services/agentpools/spec.go b/azure/services/agentpools/spec.go
index 3511a848..14d6e718 100644
--- a/azure/services/agentpools/spec.go
+++ b/azure/services/agentpools/spec.go
@@ -171,8 +171,10 @@ func (s *AgentPoolSpec) Parameters(existing interface{}) (params interface{}, er
                diff := cmp.Diff(normalizedProfile, existingProfile)
                if diff == "" {
                        // agent pool is up to date, nothing to do
+                       fmt.Print("agent pool is up to date, nothing to do\n")
                        return nil, nil
                }
+               fmt.Printf("Should merge tags here diff: %s", diff)
        }
```
And you are right there is a diff:
```
capz-controller-manager-86fd5fcfc-kwb7k manager Should merge tags here diff:   containerservice.AgentPool{
capz-controller-manager-86fd5fcfc-kwb7k manager   	Response: {},
capz-controller-manager-86fd5fcfc-kwb7k manager   	ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
capz-controller-manager-86fd5fcfc-kwb7k manager   		... // 24 identical fields
capz-controller-manager-86fd5fcfc-kwb7k manager   		ScaleSetEvictionPolicy: "",
capz-controller-manager-86fd5fcfc-kwb7k manager   		SpotMaxPrice:           nil,
capz-controller-manager-86fd5fcfc-kwb7k manager - 		Tags:                   map[string]*string{},
capz-controller-manager-86fd5fcfc-kwb7k manager + 		Tags:                   nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   		NodeLabels:             nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   		NodeTaints:             nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   		... // 7 identical fields
capz-controller-manager-86fd5fcfc-kwb7k manager   	},
capz-controller-manager-86fd5fcfc-kwb7k manager   	ID:   nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   	Name: nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   	Type: nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   }
```

Reply:
I added some tags to understand better the output:
```
---
apiVersion: infrastructure.cluster.x-k8s.io/v1beta1
kind: AzureManagedMachinePool
metadata:
  name: pool1
  namespace: default
spec:
  enableNodePublicIP: false
  mode: User
  osDiskSizeGB: 40
  sku: Standard_DC4s_v2
  additionalTags:
    owner: "one"
 ```

```
capz-controller-manager-86fd5fcfc-kwb7k manager Should merge tags here diff:   containerservice.AgentPool{
capz-controller-manager-86fd5fcfc-kwb7k manager   	Response: {},
capz-controller-manager-86fd5fcfc-kwb7k manager   	ManagedClusterAgentPoolProfileProperties: &containerservice.ManagedClusterAgentPoolProfileProperties{
capz-controller-manager-86fd5fcfc-kwb7k manager   		... // 24 identical fields
capz-controller-manager-86fd5fcfc-kwb7k manager   		ScaleSetEvictionPolicy: "",
capz-controller-manager-86fd5fcfc-kwb7k manager   		SpotMaxPrice:           nil,
capz-controller-manager-86fd5fcfc-kwb7k manager - 		Tags:                   map[string]*string{"owner": &"one"},
capz-controller-manager-86fd5fcfc-kwb7k manager + 		Tags:                   nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   		NodeLabels:             nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   		NodeTaints:             nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   		... // 7 identical fields
capz-controller-manager-86fd5fcfc-kwb7k manager   	},
capz-controller-manager-86fd5fcfc-kwb7k manager   	ID:   nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   	Name: nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   	Type: nil,
capz-controller-manager-86fd5fcfc-kwb7k manager   }
```

Reply:
I figured out the state above is transient because I just had added the Tag.

The update forever happens only when there are no tags associated with the nodepool because there is a comparison between `nil` and `map[string]*string{}`.

```
capz-controller-manager-86fd5fcfc-kwb7k manager - 		Tags:                   map[string]*string{},
capz-controller-manager-86fd5fcfc-kwb7k manager + 		Tags:                   nil,
```

Reply:
When all my nodepools have some tag I correctly see:

```
capz-controller-manager-86fd5fcfc-kwb7k manager agent pool is up to date, nothing to do
capz-controller-manager-86fd5fcfc-kwb7k manager agent pool is up to date, nothing to do
```

Reply:
@CecileRobertMichon problem comes from here:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/2349883d1effe6344690459b40e5dc26138010df/azure/converters/tags.go#L35-L41

This will always return a `map[string]*string{}` right ?

Reply:
I pushed this commit:
https://github.com/zioproto/cluster-api-provider-azure/commit/5a08d8c7aaee8025f79e45f5065a7e19e7f35142

Reply:
> This will always return a map[string]*string{} right

IMO this is a bug, it should return nil if the src is nil, could you please update that function and add a unit test to validate that behavior? That way you can drop the workaround

Reply:
For better readability and better future audit of the code changes, I think it is better to address the converters issue in a different PR. I understand if you want to hold this PR until we fix the converters. I suspect at least `TagsToMap` and `MapToTags` are broken in the same way.

Reply:
https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2802